osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs

```diff
@@ -154,10 +154,17 @@ private void load()
 
             scoreRetrievalRequest = new GetScoresRequest(fetchBeatmapInfo, fetchRuleset, Scope, requestMods);
 
-            scoreRetrievalRequest.Success += response => SetScores(
-                scoreManager.OrderByTotalScore(response.Scores.Select(s => s.ToScoreInfo(rulesets, fetchBeatmapInfo))),
-                response.UserScore?.CreateScoreInfo(rulesets, fetchBeatmapInfo)
-            );
+            scoreRetrievalRequest.Success += response => Schedule(() =>
+            {
+                // Beatmap may have changed since fetch request. Can't rely on request cancellation due to Schedule inside SetScores.
+                if (!fetchBeatmapInfo.Equals(BeatmapInfo))
```

Comment:
Now I'm not saying this isn't correct, but there are two other things in the request that could potentially trigger the same bug (ruleset, leaderboard scope, request mods) that could cause the same thing. Probably wanna handle these too if we're fixing anyways.

Reply:
I'm actually not so sure about those because the mentioned edge case was specifically where a non-online request was overriding the online retrieval too early.

But you're probably right on that it could happen, just much more rarely. This fix is pretty ugly, but I didn't want to touch the `SetScores` method due to scary online comments. Maybe it's time to restructure the whole flow to suck less..

Reply:
> because the mentioned edge case was specifically where a non-online request was overriding the online retrieval too early

That probably eliminates out ruleset as a cause, but I think leaderboard scope and mods could cause it as those are supporter-only features.

To be fair I looked at `SetScores()` and noped away. That class needs to go ~~die~~ get rewritten at this point.

I wouldn't be opposed to a local check of all 4 request params here if I'm honest. Probably wouldn't implement equality on the request itself though because equality overrides suck. Could maybe be a method (`IsEquivalent()` or similar).